protobuf_serialization/internal.nim

```diff
@@ -0,0 +1,156 @@
+#Variables needed by the Reader and Writer which should NOT be exported outside of this library.
+
+import options
+import sets
+import macros
+
+import varint
+export varint
+
+type
+  ProtobufWireType* = enum
+    VarInt, Fixed64, LengthDelimited, StartGroup, EndGroup, Fixed32
+
+  #Number types which are platform-dependent and therefore unsafe.
+  PlatformDependentTypes* = int or uint or float
+
+  #Castable length delimited types.
+  #These can be directly casted from a seq[byte] and do not require a custom converter.
+  CastableLengthDelimitedTypes* = string or seq[char or byte or uint8 or bool]
+  #This type is literally every other type.
+  #Every other type is considered custom, due to the need for their own converters.
+  #While cstring/array are built-ins, and therefore should have converters provided, but they still need converters.
+  LengthDelimitedTypes* = not (VarIntTypes or FixedTypes)
+
+proc getTypeChain(impure: NimNode): seq[NimNode] {.compileTime.} =
```

Comment:
Some documentation here would be appropriate.

As far as I can tell, it's quite possible that you don't need this at all, because you can implement `flatType` as something close to `type(flatMapInternal(value))`.

`isPotentiallyNull` will be a simple recursive function replicating the logic in `getTypeChain`, but without the error-prone type AST business.